core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java

```diff
@@ -124,16 +125,54 @@ public String generateLoadBalancerEndpoint(Stack stack) {
         }
     }
 
+    /*
+     *  follows these peferences:
+     *   public fqdn
+     *   public DNS
+     *   public IP
+     *   private FQDN
+     *   private DNS
+     *   private IP
+     */
     public String getLoadBalancerUserFacingFQDN(Long stackId) {
         Set<LoadBalancer> loadBalancers = loadBalancerPersistenceService.findByStackId(stackId).stream()
-            .filter(lb -> StringUtils.isNotEmpty(lb.getFqdn()))
-            .collect(Collectors.toSet());
-        LoadBalancer loadBalancer = loadBalancers.stream()
-            .filter(lb -> LoadBalancerType.PUBLIC.equals(lb.getType()))
-            .findAny()
-            .orElseGet(() -> loadBalancers.stream().findAny().orElse(null));
+                .filter(lb -> StringUtils.isNotEmpty(lb.getDns()) ||
+                        StringUtils.isNotEmpty(lb.getIp()) || StringUtils.isNotEmpty(lb.getFqdn()))
+                .collect(Collectors.toSet());
+        if (loadBalancers.isEmpty()) {
+            return null;
+        }
+        Iterator<LoadBalancer> iterator = loadBalancers.iterator();
+        LoadBalancer best = iterator.next();
+        while (iterator.hasNext()) {
+            LoadBalancer balancer = iterator.next();
+            if (!best.getType().equals(balancer.getType())) {
+                if (best.getType().equals(LoadBalancerType.PRIVATE)) {
+                    best = balancer;
+                }
+                continue;
+            }
+            if (StringUtils.isNotEmpty(balancer.getFqdn()) && StringUtils.isEmpty(best.getFqdn())) {
+                best = balancer;
+            }
 
-        return loadBalancer == null ? null : loadBalancer.getFqdn();
+        }
+        return getBestAddressable(best);
+
+    }
+
+    private String getBestAddressable(LoadBalancer lb) {
+        Preconditions.checkNotNull(lb);
```

Comment:
how could this happen? I don't think this is a valid check and won't be better then a nullpointer on the next line

Reply:
Unresolving this because Mihaly has a point. This is only called when `findAny()` find a valid element, so when could it be null? I understand the caution of wanting to check that it's not, but I think a better path would be to make sure we don't call this method will a null argument.

Reply:
That's a good point! I think it's also probably valid to return null in that case, which can signal to anyone else using the method that they should use a different load balancer.